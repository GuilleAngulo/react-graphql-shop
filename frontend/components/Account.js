import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

import Form from './styles/Form';
import Error from './ErrorMessage';


const UPDATE_USER_MUTATION = gql`
    mutation UPDATE_USER_MUTATION(
        $id: ID!
        $name: String
        $email: String
        $password: String
    ) {
    updateUser(id: $id, name: $name, email: $email, password: $password) {
            id
            name
            email
        }
    }
`;


class Account extends Component {
    
    static propTypes = {
        user: PropTypes.shape({
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
            email: PropTypes.string.isRequired,
        }).isRequired,
    };

    state = { 
        id: this.props.user.id,
        name: this.props.user.name,
        email: this.props.user.email,
    };

    handleChange = e => {
        const { name, value } = e.target;
        this.setState({ [name]: value });
    }

    updateUser = async (e, updateUserMutation) => {
        e.preventDefault();
        const res = await updateUserMutation({
            variables: {
                ...this.state,
            },
        });
    }

    render() {
        const { user } = this.props;
        return (
                <Mutation 
                    mutation={ UPDATE_USER_MUTATION }
                    variables={this.state}>
                        {(updateUser, { loading, error }) => user && (
                            <Form 
                            data-test="form"
                            onSubmit={e => this.updateUser(e, updateUser)}>
                                <Error error={error} />
                                <fieldset disabled={loading} aria-busy={loading}>
                                    
                                    <label htmlFor="name">
                                        Name
                                        <input 
                                            type="text" 
                                            id="name" 
                                            name="name" 
                                            placeholder="Name" 
                                            required 
                                            defaultValue={user.name}
                                            onChange={this.handleChange}
                                            />
                                    </label>

                                    <label htmlFor="email">
                                        Email
                                        <input 
                                            type="text" 
                                            id="email" 
                                            name="email" 
                                            placeholder="Email" 
                                            required 
                                            defaultValue={user.email}
                                            onChange={this.handleChange}
                                            />
                                    </label>

                                    <label htmlFor="password">
                                        Your Actual Password
                                        <input 
                                            type="password" 
                                            id="password" 
                                            name="password" 
                                            placeholder="Type your current password" 
                                            required
                                            onChange={this.handleChange}
                                            />
                                    </label>
                                    <button type="submit">Sav{loading ? 'ing' : 'e'} Changes</button>
                                </fieldset>
                            </Form>
                        )}    
                    </Mutation>
                
        );
    }
}

export default Account;
export { UPDATE_USER_MUTATION };