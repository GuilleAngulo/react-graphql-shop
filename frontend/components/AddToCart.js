import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
    mutation ADD_TO_CART_MUTATION($id: ID!) {
        addToCart(id: $id) {
            id
            quantity
            item {
                id
                image
                price
                title
                description
            }
        }
    }
`;

class AddToCart extends React.Component {
    
    static propTypes = {
        item: PropTypes.shape({
            id: PropTypes.string.isRequired,
            title: PropTypes.string.isRequired,
            price: PropTypes.number.isRequired,
            image: PropTypes.string.isRequired,
        }).isRequired,
    };

    update = (cache, payload) => {
        //1. Read the cache
        //console.log('Updating the cart.');
        const data = cache.readQuery({ query: CURRENT_USER_QUERY });
        const { cart } = data.me; 
        const cartItemId = payload.data.addToCart.id;
        //CASE 1 - Updating from optimisticResponse
        if (cartItemId.startsWith('-')) {
            //console.log('1. Optimistic Response.');
            const itemId = payload.data.addToCart.item.id;
            
            const cartItemExists = cart.findIndex(
                cartItem => cartItem.item.id === itemId
            );

            if (cartItemExists === -1)
                cart.push(payload.data.addToCart);
            else
                cart[cartItemExists].quantity += 1;
                
            cache.writeQuery({ query: CURRENT_USER_QUERY, data })
        } 
        //CASE 2 - Updating from the server
        else {
            //console.log('2. Server Update.');
            const cartItemExists = cart.findIndex(
                cartItem => cartItem.id === cartItemId
            );

            if (cartItemExists === -1) {
                cart.push(payload.data.addToCart);
            }  
            else {
                cart.splice(cartItemExists, 1, payload.data.addToCart);
            }
            cache.writeQuery({ query: CURRENT_USER_QUERY, data })
        }
    }

    render() {
        const { id } = this.props.item;
        const { item } = this.props;
        return (
            <Mutation 
            mutation={ADD_TO_CART_MUTATION}
            variables= {{ id }}
            //refetchQueries={[{ query: CURRENT_USER_QUERY }]}
            update={this.update}
            optimisticResponse={{
                __typename: 'Mutation',
                addToCart: {
                    __typename: 'CartItem',
                    id: Math.round(Math.random() * -1000000).toString(),
                    quantity: 1,
                    item,
                },
            }}
            >
                {(addToCart, { loading }) => (
                <button 
                onClick={addToCart}
                disabled={loading}>
                    Add{loading && 'ing'} to Cart ðŸ›’
                </button>
                )}
            </Mutation> 
        );
    }
}

export default AddToCart;
export { ADD_TO_CART_MUTATION };