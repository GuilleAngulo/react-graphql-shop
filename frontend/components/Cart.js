import React from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { adopt } from 'react-adopt';

import User from './User';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import CartItem from './CartItem';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import TakeMyMoney from './TakeMyMoney';

const LOCAL_STATE_QUERY = gql`
    query {
        cartOpen @client
    }
`;

const TOOGLE_CART_MUTATION = gql`
    mutation {
        toogleCart @client
    }
`;
/* eslint-disable */
const Composed = adopt({
    user: ({ render }) => <User>{render}</User> ,
    toogleCart: ({ render }) => <Mutation mutation={TOOGLE_CART_MUTATION}>{render}</Mutation>,
    localState: ({ render }) => <Query query={LOCAL_STATE_QUERY} >{render}</Query>,
});
/* eslint-enable */

function totalItems(cart) {
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0);
}

const Cart = () => (
    <Composed>
        {({ user, toogleCart, localState }) => {
            const me = user.data.me;
            if (!me) return null;
            return (
                <CartStyles open={localState.data.cartOpen}>
                    <header>
                        <CloseButton 
                        onClick={toogleCart} 
                        title="close">
                            &times;
                        </CloseButton>
                        <Supreme>{me.name}'s Cart</Supreme>
                        <p>You have {totalItems(me.cart)}  item{me.cart.length === 1 ? '' : 's'} in your cart.</p>
                    </header>
                    <ul>
                        {me.cart.map(cartItem => 
                        <CartItem 
                        key={cartItem.id} 
                        cartItem={cartItem} 
                        />)}
                    </ul>
                    <footer>
                        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                        {me.cart.length && (
                            <TakeMyMoney>
                                <SickButton>Checkout</SickButton>
                            </TakeMyMoney>
                        )}
                    </footer>
                </CartStyles>
            );
        }}
    </Composed>
);

export default Cart;
export { LOCAL_STATE_QUERY, TOOGLE_CART_MUTATION };