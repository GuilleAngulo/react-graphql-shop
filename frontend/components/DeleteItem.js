import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';
//import { PAGINATION_QUERY } from './Pagination';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
        }
    }
`;

class DeleteItem extends Component {
    update = (cache, payload) => {
        //Manually update the cache on the client, so it matches the server
        //1. Read the cache for the items we want
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
        //In the payload we have the identification of the deleted item
        //2. Filter the deleted item out of the page
        const filteredItems = data.items.filter(item => item.id !== payload.data.deleteItem.id);
        //data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
        //3. Put the items back
        cache.writeQuery({ 
            query: ALL_ITEMS_QUERY, 
            data: { items: filteredItems } 
        });
        //cache.writeQuery({ query: ALL_ITEMS_QUERY, data });

        /**PAGINATION */
        /*const pagination = cache.readQuery({ query: PAGINATION_QUERY});
        pagination.itemsConnection.aggregate.count -= 1;
        cache.writeQuery({ 
            query: PAGINATION_QUERY, 
            data: pagination,
        });*/

    };

    render() {
        return (
            <Mutation 
            mutation={DELETE_ITEM_MUTATION}
            variables={{ id: this.props.id }}
            update={this.update}
            optimisticResponse={{
                __typename: 'Mutation',
                deleteItem: {
                    __typename: 'Item',
                    id: this.props.id,
                },
            }}
            >
                {(deleteItem, { error }) => (
                    <button 
                    onClick={() => {
                        if (confirm('Are you sure you want to delete this item?')) {
                            deleteItem().catch(err => {
                                alert(err.message);
                            });
                        }
                    }}>
                        {this.props.children}
                    </button>
                )}
            </Mutation>
        );
    }
}

export default DeleteItem;