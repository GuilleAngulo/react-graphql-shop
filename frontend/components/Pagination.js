import React from 'react';
import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import { Query } from 'react-apollo';

import PaginationStyles from './styles/PaginationStyles';
import { perPage, appTitle } from '../config';
import Error from './ErrorMessage';

const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`;

const Pagination = props => (
        <Query query={PAGINATION_QUERY}>
        {({ data, loading, error }) => {
            if (loading) return <p>Loading...</p>;
            if (error) return <Error error={error} />;
            const count = data.itemsConnection.aggregate.count;
            const pages = Math.ceil(count / perPage);
            const page = props.page;
            return (
            <PaginationStyles data-test="pagination">
                <Head>
                    <title>{appTitle} - Page {page} of {pages}  </title>
                </Head>
                <Link 
                prefetch
                href={{
                    pathname: 'items',
                    query: { page: page - 1},
                }}>
                <a className="prev" aria-disabled={page <= 1}>← Prev</a>
                </Link>
                <p>
                    Page {page} of &nbsp;
                    <span className="totalPages">
                    {pages}
                    </span>
                </p>
                <p>{count} Items Total</p>
                <Link 
                prefetch
                href={{
                    pathname: 'items',
                    query: { page: page + 1},
                }}>
                <a className="next" aria-disabled={page >= pages}>Next →</a>
                </Link>
            </PaginationStyles>
            );
        }}
        </Query>
)

export default Pagination;
export { PAGINATION_QUERY };