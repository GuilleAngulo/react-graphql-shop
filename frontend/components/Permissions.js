import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

import Table from './styles/Table';
import SickButton from './styles/SickButton';
import Error from './ErrorMessage';


const UPDATE_PERMISSIONS_MUTATION = gql`
    mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userId: ID!) {
        updatePermissions(permissions: $permissions, userId: $userId) {
            id
            permissions
            name
            email
        }
    }
`;

const ALL_PERMISSIONS = gql`
    query ALL_PERMISSIONS {
        __type(name: "Permission") {
            name
            enumValues {
                name
            }
        }
    }
`;


const ALL_USERS_QUERY = gql`
    query ALL_USERS_QUERY {
        users {
            id
            name
            email
            permissions
        }
    }
`;


const Permissions = props => (
    <Query query={ALL_USERS_QUERY}>
        {({data, loading, error}) => {
        if (loading) return <p>Loading...</p>;
        if (error) return <Error error={error} />;
        return (
            <div>
                <div>
                <h2>Manage Permissions</h2>
                <Table>
                    <thead>
                        <tr>
                            <th>Name</th>
                            <th>Email</th>
                            <Query query={ALL_PERMISSIONS}>
                                {({data, loading, error}) => {
                                    if (loading) return <th>Loading...</th>;
                                    if (error) return  <th><Error error={error} /></th>;
                                    if (!data) return (<th>No permissions found.</th>);   
                                    return data.__type.enumValues.map(permission => <th key={permission.name}>{permission.name}</th>);
                                }}
                            </Query>
                            <th>ğŸ‘‡</th>
                        </tr>
                    </thead>
                    <tbody>
                        {data.users.map(user => <UserPermissions key={user.id} user={user} />)}
                    </tbody>
                </Table>
                </div>
            </div>
        );
        }}
    </Query>
);

class UserPermissions extends React.Component {
    static propTypes = {
        user: PropTypes.shape({
            name: PropTypes.string,
            email: PropTypes.string,
            id: PropTypes.string,
            permissions: PropTypes.array,
        }).isRequired,
    };

    state = {
        permissions: this.props.user.permissions,
    }

    handlePermissionChange = (e) => {
    /*In case is required to call the function inside the onChange function
    handlePermissionChange = (e, updatePermissions) => { */
        //console.log(e.target.value);
        //console.log(e.target.checked);
        const checkbox = e.target;
        //Take a copy of the current permissions
        let updatedPermissions = [...this.state.permissions];
        //Figure out if it needs to remove or add this permission
        if (checkbox.checked) {
            //Add it in
            updatedPermissions.push(checkbox.value);
        } else {
            updatedPermissions = updatedPermissions.filter(permission => permission !== checkbox.value);
        }
        this.setState({ permissions: updatedPermissions });

        /**In case wanting the onChange checkbox to update permissions
        We need to pass the function as a callback because setState is 
        asynchronous, so can be runing before the update call
        this.setState({ permissions: updatedPermissions }, function() {
            updatePermissions();
        });
        
        AT ONCHANGE SECTION 
        onChange={(e) => this.handlePermissionChange(e, updatePermissions)}
        **/
    }
    
    render() {
        const user = this.props.user;
        return (
            <Mutation mutation={UPDATE_PERMISSIONS_MUTATION}
            variables={{
                permissions: this.state.permissions,
                userId: this.props.user.id,
            }}>
                {(updatePermissions, { loading, error }) => (
                <>
                { error && <tr><td colspan="8"><Error error={error} /></td></tr> }
                <tr>
                    <td>{user.name}</td>
                    <td>{user.email}</td>
                    <Query query={ALL_PERMISSIONS}>
                        {({data, loading, error}) => {
                            if (loading) return <th>Loading...</th>;
                            if (error) return  <Error error={error} />;
                            if (!data) return (<th>No permissions found.</th>);   
                            return data.__type.enumValues.map(permission => 
                                <td key={permission.name}>
                                    <label htmlFor={`${user.id}-permission-${permission.name}`}>
                                        <input
                                        id={`${user.id}-permission-${permission.name}`} 
                                        type="checkbox" 
                                        checked={this.state.permissions.includes(permission.name)}
                                        value={permission.name}
                                        onChange={this.handlePermissionChange}/>
                                    </label>
                                </td>
                            );
                        }}
                    </Query>
                    <td>
                        <SickButton
                        type="button"
                        disabled={loading}
                        onClick={updatePermissions}>
                            Updat{loading ? 'ing' : 'e' }
                        </SickButton>
                    </td>
                </tr>
                </>
                )}
            </Mutation>
        );
    }
}

export default Permissions;