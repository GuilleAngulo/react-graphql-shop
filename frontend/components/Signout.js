import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import { CURRENT_USER_QUERY } from './User';

const SIGNOUT_MUTATION = gql`
    mutation SIGNOUT_MUTATION {
        signout {
            message
        }
    }   
`;

const Signout = props => (
    <Mutation 
    mutation={SIGNOUT_MUTATION}
    refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
        {signout => <button onClick={signout}>Sign Out</button>}
    </Mutation>
);

export default Signout;