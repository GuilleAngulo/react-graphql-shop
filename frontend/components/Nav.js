import Link from 'next/link';
import { Mutation } from 'react-apollo';
import { TOOGLE_CART_MUTATION } from '../components/Cart';

import NavStyles from './styles/NavStyles';
import User from './User';
import Signout from './Signout';
import CartCount from './CartCount';

const Nav = () => (
    <User>
        {({ data: { me } }) => (
        <NavStyles data-test="nav">
            <Link href="/items">
                <a>Shop</a>
            </Link>
            {me && (
                <>
                    <Link href="/sell">
                        <a>Sell</a>
                    </Link>
                    <Link href="/orders">
                        <a>Orders</a>
                    </Link>
                    <Link href="/me">
                        <a>Account</a>
                    </Link>
                    <Signout />
                    <Mutation mutation={TOOGLE_CART_MUTATION}>
                        {(toogleCart) => (
                            <button onClick={toogleCart}>
                                My Cart
                                <CartCount 
                                count={
                                    me.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
                                }>

                                </CartCount>
                            </button>    
                        )}
                    </Mutation>

                    {/* Permissions tab showed only for ADMIN users */}
                    {(me.permissions.findIndex(permission => permission === 'ADMIN') > - 1) && (
                        <Link href="/permissions">
                            <a>Permissions üîê</a>
                        </Link>
                    )}


                </>
            )}

            {!me && (
                <Link href="/signup">
                    <a>Sign In</a>
                </Link>
            )}
        </NavStyles>
        )}
    </User> 
);

export default Nav;