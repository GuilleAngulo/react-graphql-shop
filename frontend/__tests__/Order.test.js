import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer} from 'react-apollo';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem, fakeItem, fakeOrder } from '../lib/testUtils';
import OrderStyles from '../components/styles/OrderStyles';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';

const order = fakeOrder();

const mocks = [
    {
        request: { query: SINGLE_ORDER_QUERY, variables: { id: 'ord123' } },
        result: {
            data: {
                order: fakeOrder(),
            },
        },
    },
]

describe('<Order/>', () => {
    it('renders and matches snapshot', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <Order id="ord123"/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const order = wrapper.find('div[data-test="order"]');
        expect(toJSON(order)).toMatchSnapshot();
    });

    /**
    it('renders the item count and order total properly', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <Order id="ord123"/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        const order = wrapper.find('div[data-test="order"]');
        const orderTotal = order.find('span[data-test="order-total"]');
        const itemCount = wrapper.find('span[data-test="item-count"]');
        console.log(orderTotal.children().text());
        expect(orderTotal.children().text()).toBe(fakeOrder.total);
        expect(itemCount).toBe(fakeOrder.items.length);
    }); 
     */
});