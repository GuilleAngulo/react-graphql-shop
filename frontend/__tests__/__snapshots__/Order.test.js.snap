// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order/> renders and matches snapshot 1`] = `
<div
  className="OrderStyles-k2yeg-0 bwaDUN"
  data-test="order"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <p>
    <span>
      Order ID: 
    </span>
    <span>
      ord123
    </span>
  </p>
  <p>
    <span>
      Charge 
    </span>
    <span>
      ch_123
    </span>
  </p>
  <p>
    <span>
      Date 
    </span>
    <span>
      March 31, 2018 9:00 PM
    </span>
  </p>
  <p>
    <span
      data-test="order-total"
    >
      Order 
    </span>
    <span>
      $400
    </span>
  </p>
  <p>
    <span>
      Item Count 
    </span>
    <span
      data-test="item-count"
    >
      2
    </span>
  </p>
  <div
    className="items"
  >
    <div
      className="order-item"
      key="b4505a74-1f67-4ee1-9558-e56760116bcd"
    >
      <img
        alt="nihil quam nulla est unde sit ex"
        src="dolorem.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          nihil quam nulla est unde sit ex
        </h2>
        <p>
          Qty: 
          1
        </p>
        <p>
          Each: 
          $42.34
        </p>
        <p>
          Subtotal: 
          $42.34
        </p>
        <p>
          expedita est doloremque explicabo eaque necessitatibus sit
        </p>
      </div>
    </div>
    <div
      className="order-item"
      key="4cef4042-dadc-4ad4-a066-08d01ced82d4"
    >
      <img
        alt="voluptas culpa et vero repellat eveniet assumenda"
        src="et.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          voluptas culpa et vero repellat eveniet assumenda
        </h2>
        <p>
          Qty: 
          1
        </p>
        <p>
          Each: 
          $42.34
        </p>
        <p>
          Subtotal: 
          $42.34
        </p>
        <p>
          quis quae iusto sapiente nostrum aut corrupti
        </p>
      </div>
    </div>
  </div>
</div>
`;
